File: app/src/main/java/com/security/passwordmanager/data/DataDao.java
package com.security.passwordmanager.data;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import java.util.List;

@Dao
public abstract class DataDao {

    @Insert(entity = Website.class, onConflict = OnConflictStrategy.REPLACE)
    public abstract void addWebsite(Website website);

    @Insert(entity = BankCard.class, onConflict = OnConflictStrategy.REPLACE)
    public abstract void addBankCard(BankCard bankCard);

    @Update(entity = Website.class, onConflict = OnConflictStrategy.REPLACE)
    public abstract void updateWebsite(Website website);

    @Update(entity = BankCard.class, onConflict = OnConflictStrategy.REPLACE)
    public abstract void updateBankCard(BankCard bankCard);

    @Query("SELECT * FROM WebsiteTable ORDER BY nameWebsite ASC")
    public abstract List<Website> getWebsiteList();

    @Query("SELECT * FROM BankTable ORDER BY bankName ASC")
    public abstract List<BankCard> getBankCardList();

    @Query("SELECT * FROM WebsiteTable WHERE address = :address")
    public abstract List<Website> getAccountList(String address);

    @Query("SELECT * FROM BankTable WHERE bankName = :bankName")
    public abstract List<BankCard> getBankAccountList(String bankName);

    @Query("SELECT * FROM WebsiteTable " +
            "WHERE nameWebsite LIKE '%' || :query || '%' OR " +
            "address LIKE '%' || :query || '%' OR " +
            "nameAccount LIKE '%' || :query || '%' " +
            "ORDER BY nameWebsite ASC")
    public abstract List<Website> searchWebsite(String query);

    @Query("SELECT * FROM BankTable " +
            "WHERE bankName LIKE '%' || :query || '%' " +
            "ORDER BY bankName")
    public abstract List<BankCard> searchBankCard(String query);

    @Delete(entity = Website.class)
    public abstract void deleteWebsite(Website website);

    @Delete(entity = BankCard.class)
    public abstract void deleteBankCard(BankCard bankCard);

    @Query("DELETE FROM WebsiteTable WHERE address = :url")
    public abstract void deleteWebsite(String url);

    @Query("DELETE FROM BankTable WHERE bankName = :bankName")
    public abstract void deleteBankCard(String bankName);

    @Transaction
    public void search(String query) {
        searchWebsite(query);
        searchBankCard(query);
    }
}

File: app/src/main/java/com/security/passwordmanager/data/DataDao.kt
package com.security.passwordmanager.data

import androidx.room.*
import com.security.passwordmanager.Pair
import com.security.passwordmanager.model.BankCard
import com.security.passwordmanager.model.Website

@Dao
abstract class DataDao {

    /**
     * Добавление в базу
     */
    @Insert(entity = Website::class, onConflict = OnConflictStrategy.REPLACE)
    abstract fun addWebsite(website: Website)

    @Insert(entity = BankCard::class, onConflict = OnConflictStrategy.REPLACE)
    abstract fun addBankCard(bankCard: BankCard)


    /**
     * Обновление элементов
     */
    @Update(entity = Website::class, onConflict = OnConflictStrategy.REPLACE)
    abstract fun updateWebsite(website: Website)

    @Update(entity = BankCard::class, onConflict = OnConflictStrategy.REPLACE)
    abstract fun updateBankCard(bankCard: BankCard)


    /**
     * Получение всех элементов
     */
    @Query("SELECT * FROM WebsiteTable WHERE email = :email ORDER BY nameWebsite ASC")
    abstract suspend fun getWebsiteList(email: String): MutableList<Website>

    @Query("SELECT * FROM BankTable WHERE email = :email ORDER BY bankName ASC")
    abstract suspend fun getBankCardList(email: String): MutableList<BankCard>


    /**
     * Поиск
     */
    @Query("SELECT * FROM WebsiteTable " +
            "WHERE email = :email AND (" +
            "nameWebsite LIKE '%' || :query || '%' OR " +
            "address LIKE '%' || :query || '%' OR " +
            "nameAccount LIKE '%' || :query || '%') " +
            "ORDER BY nameWebsite ASC")
    abstract suspend fun searchWebsite(email: String, query: String): MutableList<Website>

    @Query("SELECT * FROM BankTable " +
            "WHERE email = :email AND (" +
            "bankName LIKE '%' || :query || '%' OR " +
            "bankCardName LIKE '%' || :query || '%') " +
            "ORDER BY bankName")
    abstract suspend fun searchBankCard(email: String, query: String): MutableList<BankCard>


    /**
     * Удаление одного элемента
     */
    @Delete(entity = Website::class)
    abstract fun deleteWebsite(website: Website)

    @Delete(entity = BankCard::class)
    abstract fun deleteBankCard(bankCard: BankCard)


    /**
     * удаление нескольких элементов по ключу
     */
    @Query("DELETE FROM WebsiteTable WHERE email = :email AND address = :url")
    abstract fun deleteWebsite(email: String, url: String)

    @Query("DELETE FROM BankTable WHERE email = :email AND bankName = :bankName")
    abstract fun deleteBankCard(email: String, bankName: String)


    @Transaction
    open suspend fun search(email: String, query: String) =
        Pair(searchWebsite(email, query), searchBankCard(email, query))
}